crates/scmi/src/scmi.rs

```diff
@@ -0,0 +1,1377 @@
+//! devices), see [crate::devices] modules.
+
+use std::{
+    cmp::min,
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use itertools::Itertools;
+use log::{debug, error, info, warn};
+use thiserror::Error as ThisError;
+
+pub(crate) type MessageHeader = u32;
+
+pub(crate) const MAX_SIMPLE_STRING_LENGTH: usize = 16; // incl. NULL terminator
+
+/// Wrapper around SCMI values of the basic types SCMI defines.
+///
+/// Everything communicating to/from SCMI must be composed of them.
+// SCMI specification talks about Le32 parameter and return values.
+// VirtIO SCMI specification talks about u8 SCMI values.
+// Let's stick with SCMI specification for implementation simplicity.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) enum MessageValue {
+    Signed(i32),
+    Unsigned(u32),
+    String(String, usize), // string, expected characters
+}
+
+impl MessageValue {
+    pub(crate) fn get_unsigned(&self) -> u32 {
+        match self {
+            MessageValue::Unsigned(value) => *value,
+            _ => panic!("Wrong parameter"),
+        }
+    }
+}
+
+pub(crate) type MessageValues = Vec<MessageValue>;
+
+/// Enumeration of SCMI message types, mapped to the corresponding SCMI codes.
+///
+/// The only one we currently support is [MessageType::Command].
+#[derive(Debug, PartialEq)]
+enum MessageType {
+    // 4-bit unsigned integer
+    Command,     // 0
+    Unsupported, // anything else
+}
+pub(crate) type MessageId = u8;
+pub(crate) type ProtocolId = u8;
+type NParameters = u8;
+
+/// Mapping of return values to SCMI return status codes.
+#[derive(Clone, Copy)]
+#[allow(dead_code)]
+enum ReturnStatus {
+    // 32-bit signed integer
+    Success = 0,
+    NotSupported = -1,
+    InvalidParameters = -2,
+    Denied = -3,
+    NotFound = -4,
+    OutOfRange = -5,
+    Busy = -6,
+    CommsError = -7,
+    GenericError = -8,
+    HardwareError = -9,
+    ProtocolError = -10,
+    // -11..-127: reserved
+    // <-127: vendor specific
+}
+
+impl ReturnStatus {
+    fn as_value(&self) -> MessageValue {
+        MessageValue::Signed(*self as i32)
+    }
+}
+
+/// Representation of [MessageValue] sequence used to construct [ScmiResponse].
+///
+/// The sequence includes the response code (see the helper constructors for
+/// adding them) but it doesn't include the SCMI message header.  The header is
+/// added in [ScmiResponse].
+struct Response {
+    values: MessageValues,
+}
+
+impl Response {
+    fn from_status(status: ReturnStatus) -> Response {
+        Response {
+            values: vec![status.as_value()],
+        }
+    }
+
+    fn from_value(value: MessageValue) -> Response {
+        Response {
+            values: vec![ReturnStatus::Success.as_value(), value],
+        }
+    }
+
+    fn from_values(values: &mut MessageValues) -> Response {
+        let mut response_values = vec![ReturnStatus::Success.as_value()];
+        response_values.append(values);
+        Response {
+            values: response_values,
+        }
+    }
```

Comment:
```suggestion
    fn from_status(status: ReturnStatus) -> Self {
        Self {
            values: vec![status.as_value()],
        }
    }

    fn from_value(value: MessageValue) -> Self {
        Self {
            values: vec![ReturnStatus::Success.as_value(), value],
        }
    }

    fn from_values(values: &mut MessageValues) -> Self {
        let mut response_values = vec![ReturnStatus::Success.as_value()];
        response_values.append(values);
        Self {
            values: response_values,
        }
    }
```

Reply:
But all of theses sounds like From trait implementation candidates

Reply:
Right, changed to From<>.